dotnet/src/SemanticKernel/Orchestration/Plan.cs

```diff
@@ -0,0 +1,281 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.Text.Json.Serialization;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+using Microsoft.SemanticKernel.AI.TextCompletion;
+using Microsoft.SemanticKernel.SkillDefinition;
+
+namespace Microsoft.SemanticKernel.Orchestration;
+
+/// <summary>
+/// Standard Semantic Kernel callable plan.
+/// Plan is used to create trees of <see cref="ISKFunction"/>s.
+/// </summary>
+public sealed class Plan : ISKFunction
+{
+    /// <summary>
+    /// State of the plan
+    /// </summary>
+    [JsonPropertyName("state")]
+    public ContextVariables State { get; } = new();
+
+    /// <summary>
+    /// Steps of the plan
+    /// </summary>
+    [JsonPropertyName("steps")]
+    internal List<Plan> Steps { get; } = new();
+
+    /// <summary>
+    /// Named parameters for the function
+    /// </summary>
+    [JsonPropertyName("named_parameters")]
+    public ContextVariables NamedParameters { get; set; } = new();
+
+    public bool HasNextStep => this.NextStep < this.Steps.Count;
+
+    public int NextStep { get; private set; } = 0;
+
+    #region ISKFunction implementation
+
+    /// <inheritdoc/>
+    [JsonPropertyName("name")]
+    public string Name { get; set; } = string.Empty;
+
+    /// <inheritdoc/>
+    [JsonPropertyName("skill_name")]
+    public string SkillName { get; set; } = string.Empty;
+
+    /// <inheritdoc/>
+    [JsonPropertyName("description")]
+    public string Description { get; set; } = string.Empty;
+
+    /// <inheritdoc/>
+    [JsonPropertyName("is_semantic")]
+    public bool IsSemantic { get; internal set; } = false;
+
+    /// <inheritdoc/>
+    [JsonPropertyName("request_settings")]
+    public CompleteRequestSettings RequestSettings { get; internal set; } = new();
+
+    #endregion ISKFunction implementation
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="Plan"/> class with a goal description.
+    /// </summary>
+    /// <param name="goal">The goal of the plan used as description.</param>
+    public Plan(string goal)
+    {
+        this.Description = goal;
+        this.SkillName = this.GetType().FullName;
+        this.Name = goal;
```

Comment:
I would keep the Property name as `Goal`. 

We can have Plans within Plans, and potentially have names for thos different to achieve a same goal...

Reply:
The "Name" property is part of the ISKFunction implementation